lib/ansible/plugins/action/set_fact.py

```diff
@@ -51,7 +51,8 @@ def run(self, tmp=None, task_vars=None):
                                      "letters, numbers and underscores." % k)
                     return result
 
-                if not C.DEFAULT_JINJA2_NATIVE and isinstance(v, string_types) and v.lower() in ('true', 'false', 'yes', 'no'):
+                v_is_boolean = isinstance(v, string_types) and v.lower() in ('true', 'false', 'yes', 'no', 'y', 'n', 'on', 'off')
+                if v_is_boolean and not C.DEFAULT_JINJA2_NATIVE:
```

Comment:
`import ansible.module_utils.parsing.convert_bool import BOOLEANS` and use that constant

Reply:
This should just force strict and leave unchanged on typeerror
```diff --git a/lib/ansible/plugins/action/set_fact.py b/lib/ansible/plugins/action/set_fact.py
index d7fe573c1a..934245d07c 100644
--- a/lib/ansible/plugins/action/set_fact.py
+++ b/lib/ansible/plugins/action/set_fact.py
@@ -51,8 +51,11 @@ class ActionModule(ActionBase):
                                      "letters, numbers and underscores." % k)
                     return result

-                if not C.DEFAULT_JINJA2_NATIVE and isinstance(v, string_types) and v.lower() in ('true', 'false', 'yes', 'no'):
-                    v = boolean(v, strict=False)
+                if not C.DEFAULT_JINJA2_NATIVE and isinstance(v, string_types):
+                    try:
+                        v = boolean(v, strict=True)
+                    except TypeError:
+                        pass  # not valid bool string
                 facts[k] = v

         result['changed'] = False```